File: CPP/actions/action_tutorial/src/simple_action_server.cpp
/**
 * @file simple_action_server.cpp
 *
 * @brief A basic ROS2 action server node that concatenates a string 
 *         based on the number of string concatenation sent by a client within a goal request
 *
 * @author Antonio Mauro Galiano
 * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
 *
 */


#include "custom_action/action/concatenate.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"


using Concatenate = custom_action::action::Concatenate;
using GoalHandleConcatenate = rclcpp_action::ServerGoalHandle<Concatenate>;


// response to the client goal request
rclcpp_action::GoalResponse
  HandleGoal(const rclcpp_action::GoalUUID &uuid,
            std::shared_ptr<const Concatenate::Goal> goal) 
{
  RCLCPP_INFO(rclcpp::get_logger("server"),
              "Got goal request with %d string concatenations",
              goal->num_concatenations);
  // rclcpp_action::GoalUUID is an array type
  // is it the id of the goal??? Its not so clear
  (void)uuid;
  
  // conditional to reject numbers of concatenations
  if ((goal->num_concatenations > 10) && (goal->num_concatenations < 2)) 
  {
    return rclcpp_action::GoalResponse::REJECT;
  }
  return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
}

// response to cancel requests from the client
rclcpp_action::CancelResponse HandleCancel(const std::shared_ptr<GoalHandleConcatenate> goalHandle) 
{
  RCLCPP_INFO(rclcpp::get_logger("server"), "Got request to cancel goal");
  (void)goalHandle;
  return rclcpp_action::CancelResponse::ACCEPT;
}


// the goal is executed
void execute(const std::shared_ptr<GoalHandleConcatenate> goalHandle) 
{
  RCLCPP_INFO(rclcpp::get_logger("server"), "Executing the concatenation");
  rclcpp::Rate loop_rate(1);
  const auto goal = goalHandle->get_goal();
  auto feedback = std::make_shared<Concatenate::Feedback>();
  std::string myString = "HELLOWORLD";
  auto &concatenation = feedback->partial_concatenation;
  concatenation = myString;
  concatenation = concatenation + " " + myString;
  auto result = std::make_shared<Concatenate::Result>();

  for (int i = 1; (i < goal->num_concatenations) && rclcpp::ok(); ++i)
  {
    // check if there is a cancel request
    if (goalHandle->is_canceling())
    {
      result->final_concatenation = concatenation;
      goalHandle->canceled(result);
      RCLCPP_INFO(rclcpp::get_logger("server"), "Goal Canceled");
      return;
    }
    // update the final concatenation
    concatenation = concatenation + " " + myString;
    // update and publish feedback of the partial concatenation
    goalHandle->publish_feedback(feedback);
    RCLCPP_INFO(rclcpp::get_logger("server"), "Publish Feedback");

    loop_rate.sleep();
  }

  // check if goal is done
  if (rclcpp::ok())
  {
    result->final_concatenation = concatenation;
    goalHandle->succeed(result);
    RCLCPP_INFO(rclcpp::get_logger("server"), "Goal Succeeded");
  }
}

void HandleAccepted(const std::shared_ptr<GoalHandleConcatenate> goalHandle)
{
  // this needs to return quickly to avoid blocking the executor
  // so spin up a new thread
  std::thread{execute, goalHandle}.detach();
}

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("simple_action_server");

  // create an action server with three callbacks
  // 'handle_goal' and 'handle_cancel' are called by the Executor
  // (rclcpp::spin) 'execute' is called whenever 'handle_goal' returns by
  // accepting a goal
  // Calls to 'execute' are made in an available thread from a pool of four.
  auto action_server = rclcpp_action::create_server<Concatenate>(
      node, "concatenation", HandleGoal, HandleCancel, HandleAccepted);

  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: PYTHON/actions/action_tutorial_py/action_tutorial_py/simple_action_server.py
#  *
#  * @file simple_action_server.py
#  *
#  * @brief A basic ROS2 action server node that concatenates a string 
#  *         based on the number of string concatenation sent by a client within a goal request
#  *
#  * @author Antonio Mauro Galiano
#  * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
#  *
#

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse
from custom_action.action import Concatenate
import threading


class SimpleActionServer(Node):
    """Simple action server node."""

    def __init__(self):
        super().__init__('simple_action_server')
        self._action_server = ActionServer(
            self,
            Concatenate,
            'concatenation',
            self.execute_callback,
            goal_callback=self.handle_goal,
            cancel_callback=self.handle_cancel
        )

    def handle_goal(self, goal_request):
        """
        Response to the client goal request.
        """
        self.get_logger().info(
            f'Got goal request with {goal_request.num_concatenations} string concatenations'
        )
        # Conditional to reject numbers of concatenations
        if (goal_request.num_concatenations > 10) and (goal_request.num_concatenations < 2):
            return GoalResponse.REJECT
        # In ROS 2 Jazzy Python, use ACCEPT instead of ACCEPT_AND_EXECUTE
        return GoalResponse.ACCEPT

    def handle_cancel(self, goal_handle):
        """
        Response to cancel requests from the client.
        """
        self.get_logger().info('Got request to cancel goal')
        return rclpy.action.CancelResponse.ACCEPT

    def execute_callback(self, goal_handle):
        """
        The goal is executed.
        """
        self.get_logger().info('Executing the concatenation')
        
        goal = goal_handle.request
        feedback_msg = Concatenate.Feedback()
        my_string = "HELLOWORLD"
        concatenation = my_string
        concatenation = concatenation + " " + my_string
        result = Concatenate.Result()
        
        for i in range(1, goal.num_concatenations):
            if not rclpy.ok():
                break
            
            # Check if there is a cancel request
            if goal_handle.is_cancel_requested:
                result.final_concatenation = concatenation
                goal_handle.canceled()
                self.get_logger().info('Goal Canceled')
                return result
            
            # Update the final concatenation
            concatenation = concatenation + " " + my_string
            # Update and publish feedback of the partial concatenation
            feedback_msg.partial_concatenation = concatenation
            goal_handle.publish_feedback(feedback_msg)
            self.get_logger().info('Publish Feedback')
            
            rclpy.spin_once(self, timeout_sec=1.0)
        
        # Check if goal is done
        result.final_concatenation = concatenation
        if rclpy.ok():
            goal_handle.succeed()
            self.get_logger().info('Goal Succeeded')
        return result


def main(args=None):
    rclpy.init(args=args)
    action_server = SimpleActionServer()
    
    # Create an action server with three callbacks
    # 'handle_goal' and 'handle_cancel' are called by the Executor (rclpy.spin)
    # 'execute' is called whenever 'handle_goal' returns by accepting a goal
    # Note: In Python, the execute callback runs in a separate thread automatically
    
    try:
        rclpy.spin(action_server)
    except KeyboardInterrupt:
        if rclpy.ok():
            action_server.get_logger().info("Shutting down action server...")
    finally:
        action_server.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

